### PR TITLE
[dogstatsd] capture/replay: adding tagger state and pidmap support

### PR DESCRIPTION
**Pull request by truthbk**
### What does this PR do?

This PR addresses the need to capture not only the incoming traffic but also the state of the tagger through-out the traffic collection process. Naturally, there are two aspects to this PR:
- **capture**: capturing the tagger state and pid map which is required to associate the traffic's source processes with the corresponding entity (ie. container). This is done by building a map of the incoming UDS credentials, and mapping them to the entity id. When the capture is over, we iterate this pid map and store the relevant entities as the tagger state. Both the entities in the tagger, and the pid map will be stored in the file.
- **replay**: when the time to replay the traffic comes, we will load the pid map and the tagger state, and via the API we will let the agent know there's a traffic replay in progress and we will submit the tagger state and the pid map to the target agent. Because the unix socket might get a mix of replayed and live traffic we will not simply override the tagger state, instead we will add a new tagger data structure which we will hit first trying to find relevant entities, same goes for the pidmap. If nothing is found, we fallback to the usual mechanism and treat the packet as live traffic. At the end of the replay, when all the traffic has been processed, we will reset the replay tagger state and pid map back to nil, and life will continue as usual for the dogstatsd server and tagger services.

Additionally, because of the way in which I wanted to store the state and reuse previously defined protobuf messages, it became obvious that we could likely benefit greatly from a refactor in the way we managed protobuf in the repository. Rather than having our protobuf definitions littered all over the `datadog-agent` repo, it made more sense to have them all centralized in a single location. I created `pkg/proto` and dropped the definitions there such that this would help us reuse messages in a more straight-forward way, all while helping us maintain some degree of protobuf sanity in general. Also, to ensure developers can have a consistent experience when working with protobuf, grpc, etc.... the tools were pinned and added to the `inv install-tools` task to help contributors' developer experience.     

Important note: because UDS credentials are verified by the kernel, ancillary messages cannot contain arbitrary control messages. For a regular user, credentials *MUST* match the submitting process PID and UID, this means we can't impersonate another process. A superuser can send credentials for a different PID but there's a caveat as well, as verified by the kernel the PID must be a live PID on the box, otherwise you get an `ESRCH` error. We obviously cannot guarantee this in a replay situation, so *as a superuser* we set a special `GUID` for the group id, and send the PID as the UID (we are not really interested in GID or UID for origin detection). This hack allows us to replay the code and send the relevant ancillary data information with minimal processing overhead. Unfortunately there aren't many alternatives to this approach.

### Motivation

We want to be able to a) help troubleshoot UDS scenarios internally and with our customers, and b) be able to consistently replay specific traffic flows and workloads to help us benchmark and improve the dogstatsd server performance. 

### Additional Notes

This PR is large, and should likely be split into smaller PRs. Unfortunately the git history (my bad) does not lend itself well to achieve this easily. Please let me know if this is unbearable for review and I will try to help ease the burden. I did not anticipate the eventual size the PR would reach.

#### Performance Overhead

As can be expected, this feature will add overhead to the dogstatsd pipeline packet processing. It's hard to gauge the end-to-end impact via unit/benchmark tests but at the very least we can see the impact added by the packet pool manager:

```
BenchmarkPoolManagerPassthru-4          33876208                33.9 ns/op
BenchmarkPoolManagerNoPassthru-4        14455808                83.9 ns/op
BenchmarkSyncPool-4                     38432956                32.7 ns/op
```

### Describe how to test your changes

To test this feature try the following:

1. generate UDS traffic with origin detection - this requires submitting traffic from a container to the unix socket.
   - assuming you have the dogstatsd UDS socket in `/var/run/datadog-agent/statsd.sock` and `origin_detection` set to true in the config file.
   - start a docker image with: `docker run -it -v /var/run/datadog-agent/statsd.sock:/var/run/statsd.sock --rm ubuntu:latest bash`, we will use this container to generate the traffic.
   - inside the docker container, install a few pieces that will be helpful for testing: 
     - `apt-get update && apt-get install python3 python3-pip`
     - `pip3 install datadog`
     - drop this script below in the container, say in `~/generate.py`
     - now start it: `python3 ~/generate.py`
```
from datadog import initialize, statsd
import time

options = {
    "statsd_socket_path": "/var/run/statsd.sock",
}

initialize(**options)

while True:
    statsd.gauge('testing.uds.capture.gauge', 7)
    time.sleep(0.5)
```
     

2. Once the generated traffic has begun submitting, on the agent side, try capturing a lengthy trace - about 300s seconds worth of traffic with `sudo -u dd-agent -- /opt/datadog-agent/bin/datadog-agent dogstatsd-capture -d 300s`
3. Make sure you stop the traffic generating script above. 
4. Now replay back the traffic captured with `sudo -u dd-agent -- /opt/datadog-agent/bin/datadog-agent dogstatsd-replay -f <captured_file_here>`
you should see the replayed traffic generated in (1) above appear in the application with the origin detection tags (`short_image:`, etc...).

### CODE REVIEW

File: internal/tools/tools.go [hunk 1]
```diff
@@ -15,17 +15,24 @@ import (
...
 	_ "github.com/go-enry/go-license-detector/v4/cmd/license-detector"
+	_ "github.com/golang/mock/mockgen"
+	_ "github.com/golang/protobuf/protoc-gen-go"
 	_ "github.com/golangci/golangci-lint/cmd/golangci-lint"
 	_ "github.com/gordonklaus/ineffassign"
 	_ "github.com/goware/modvendor"
+	_ "github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway"
 	_ "github.com/mgechev/revive"
+	_ "github.com/stormcat24/protodep"
+	_ "github.com/stretchr/testify"
```
Comment (mx-psi) [inline internal/tools/tools.go:26]: What do we need testify here for? I don't think we need it
```suggestion
```

File: internal/tools/tools.go [hunk 2]
```diff
@@ -15,17 +15,24 @@ import (
...
+	_ "golang.org/x/lint/golint"
```
Comment (mx-psi) [inline internal/tools/tools.go:27]: We should not use golint, it's deprecated and it was removed on #8160
```suggestion
```

File: internal/tools/go.mod [hunk 3]
```diff
@@ -7,13 +7,20 @@ require (
...
+	github.com/golang/mock v1.4.4
+	github.com/golang/protobuf v1.4.2
 	github.com/golangci/golangci-lint v1.40.1
 	github.com/gordonklaus/ineffassign v0.0.0-20210225214923-2e10b2664254
 	github.com/goware/modvendor v0.3.0
+	github.com/grpc-ecosystem/grpc-gateway v1.12.1
 	github.com/mgechev/revive v1.0.6
+	github.com/stormcat24/protodep v0.0.0-20210106175019-c95189bc461b
+	github.com/stretchr/testify v1.7.0
+	golang.org/x/lint v0.0.0-20200302205851-738671d3881b
```
Comment (mx-psi) [inline internal/tools/go.mod:19]:
```suggestion
```

File: tasks/test.py [hunk 4]
```diff
@@ -42,6 +55,12 @@ def ensure_bytes(s):
     'github.com/gordonklaus/ineffassign',
     'github.com/goware/modvendor',
     'github.com/mgechev/revive',
+    'github.com/stormcat24/protodep',
+    'github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway@v1.12.2',
+    'github.com/golang/protobuf/protoc-gen-go@v1.3.2',
+    'github.com/golang/protobuf/protoc-gen-go@v1.3.2',
+    'github.com/golang/mock/mockgen@v1.5.0',
+    'golang.org/x/lint/golint',
```
Comment (mx-psi) [inline tasks/test.py:63]:
```suggestion
```

File: tasks/test.py [hunk 5]
```diff
@@ -42,6 +55,12 @@ def ensure_bytes(s):
     'github.com/gordonklaus/ineffassign',
     'github.com/goware/modvendor',
     'github.com/mgechev/revive',
+    'github.com/stormcat24/protodep',
+    'github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway@v1.12.2',
+    'github.com/golang/protobuf/protoc-gen-go@v1.3.2',
+    'github.com/golang/protobuf/protoc-gen-go@v1.3.2',
+    'github.com/golang/mock/mockgen@v1.5.0',
```
Comment (mx-psi) [inline tasks/test.py:62]: versions should not be pinned here but in the `internal/tools/go.mod` file.
```suggestion
    'github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway',
    'github.com/golang/protobuf/protoc-gen-go',
    'github.com/golang/mock/mockgen',
```

File: tasks/test.py [hunk 6]
```diff
@@ -50,9 +69,10 @@ def ensure_bytes(s):
 @task
 def install_tools(ctx):
     """Install all Go tools for testing."""
-    with ctx.cd("internal/tools"):
-        for tool in TOOL_LIST:
-            ctx.run("go install {}".format(tool))
+    with environ({'GO111MODULE': 'on'}):
```
Comment (mx-psi) [inline tasks/test.py:72]: Why do we need this? I think this should be done on a separate PR, it does not look related to the rest of changes in this PR

File: internal/tools/tools.go [hunk 7]
```diff
@@ -15,17 +15,24 @@ import (
 	_ "github.com/frapposelli/wwhrd"
```
Comment (mx-psi) [inline internal/tools/tools.go:15]: Do we really need all the new tools? I only see references to `protoc` on the tasks, can you clarify?

File: cmd/agent/api/grpc.go [hunk 8]
```diff
@@ -1,7 +1,7 @@
 // Unless explicitly stated otherwise all files in this repository are licensed
 // under the Apache License Version 2.0.
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
-// Copyright 2016-present Datadog, Inc.
+// Copyright 2016-2020 Datadog, Inc.
```
Comment (mx-psi) [inline cmd/agent/api/grpc.go:4]: This reverts #7332, we should not do it
```suggestion
// Copyright 2016-present Datadog, Inc.
```

File: go.mod [hunk 9]
```diff
@@ -189,6 +193,7 @@ require (
 	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
 	gopkg.in/zorkian/go-datadog-api.v2 v2.29.0
 	gotest.tools v2.2.0+incompatible
+	honnef.co/go/tools v0.1.1 // indirect
```
Comment (mx-psi) [inline go.mod:196]: There are a lot of `// indirect` dependencies added here that I don't think we want, can you double check?

**Review by mx-psi — DISMISSED** [review]

File: pkg/tagger/global.go [hunk 10]
```diff
@@ -73,13 +105,25 @@ func Tag(entity string, cardinality collectors.TagCardinality) ([]string, error)
 // cardinality (with tags about individual containers), or at orchestrator
 // cardinality (pod/task level).
 func TagBuilder(entity string, cardinality collectors.TagCardinality, tb *util.TagsBuilder) error {
+	//TODO: defer unlock once performance overhead of defer is negligible
```
Comment (vickenty) [inline pkg/tagger/global.go:108]: If performance is this critical, an `atomic.Value` with a func returning a tagger instance or nil might be even faster.

**Comment by mx-psi** [comment]
I moved over my comments to the new PR #8201

File: pkg/dogstatsd/packets/buffer.go [hunk 11]
```diff
@@ -8,6 +8,15 @@ package packets
...
 	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/telemetry"
+)
+
+var (
+	tlmBufferFlushedTimer = telemetry.NewCounter("dogstatsd", "packets_buffer_flush_timer",
+		nil, "Count of packets buffer flush triggered by the timer")
+	tlmBufferFlushedFull = telemetry.NewCounter("dogstatsd", "packets_buffer_flush_full",
+		nil, "Count of packets buffer flush triggered because the buffer is full")
```
Comment (hush-hush) [inline pkg/dogstatsd/packets/buffer.go:19]: Could we move the telemetry to `pkg/dogstatsd/packets/telemetry.go` ?

File: pkg/dogstatsd/packets/pool.go [hunk 12]
```diff
@@ -8,9 +8,19 @@ package packets
...
 
+var (
+	tlmPoolGet = telemetry.NewCounter("dogstatsd", "packet_pool_get",
+		nil, "Count of get done in the packet pool")
+	tlmPoolPut = telemetry.NewCounter("dogstatsd", "packet_pool_put",
+		nil, "Count of put done in the packet pool")
+	tlmPool = telemetry.NewGauge("dogstatsd", "packet_pool",
+		nil, "Usage of the packet pool in dogstatsd")
+)
+
```
Comment (hush-hush) [inline pkg/dogstatsd/packets/pool.go:23]: Same here

File: pkg/dogstatsd/packets/pool_manager.go [hunk 13]
```diff
@@ -48,19 +51,30 @@ func (p *PoolManager) Put(x interface{}) {
...
+	switch t := x.(type) {
+	case *[]byte:
+		ref = unsafe.Pointer(t)
+	case *Packet:
+		ref = unsafe.Pointer(t)
+	default:
+		log.Debugf("Unsupported type by pool manager")
+		return
+	}
+
```
Comment (hush-hush) [inline pkg/dogstatsd/packets/pool_manager.go:64]: Do we need this ? The packet pool is heavily used by dsd and type inference is costly operation. I don't think it's worth is if the only goal is to detect unsupported type.

File: pkg/dogstatsd/packets/telemetry.go [hunk 14]
```diff
@@ -11,21 +11,9 @@ import (
...
 		nil, "Number of packets in the packets channel")
 
-	// packet pool
-	tlmPoolGet = telemetry.NewCounter("dogstatsd", "packet_pool_get",
-		nil, "Count of get done in the packet pool")
-	tlmPoolPut = telemetry.NewCounter("dogstatsd", "packet_pool_put",
-		nil, "Count of put done in the packet pool")
-	tlmPool = telemetry.NewGauge("dogstatsd", "packet_pool",
-		nil, "Usage of the packet pool in dogstatsd")
-
```
Comment (hush-hush) [inline pkg/dogstatsd/packets/telemetry.go:28]: Why ?

File: pkg/dogstatsd/replay/capture.go [hunk 15]
```diff
@@ -14,6 +14,11 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/config"
 )
 
+const (
+	// GUID will be used as the GUID during capture replays
+	GUID = 999888777
```
Comment (hush-hush) [inline pkg/dogstatsd/replay/capture.go:19]: Can we expand the comment to explain why this number was chosen ?

File: pkg/dogstatsd/replay/reader.go [hunk 16]
```diff
@@ -77,12 +91,19 @@ func (tc *TrafficCaptureReader) Read() {
...
+				util.Wait(time.Second * time.Duration(msg.Timestamp-tc.last))
 			}
 		}
 
 		tc.last = msg.Timestamp
 		tc.Traffic <- msg
+
+		select {
+		case <-tc.fuse:
+			break iterate
```
Comment (hush-hush) [inline pkg/dogstatsd/replay/reader.go:103]: Why not a return ?

File: pkg/dogstatsd/replay/state.go [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+// ContainerIDForPID returns the matching container id for a pid, or an error if not found.
+func ContainerIDForPID(pid int32) (string, error) {
+	mux.RLock()
+	defer mux.RUnlock()
+
+	if pidMap == nil {
+		return "", errPidMapUnavailable
+	}
```
Comment (hush-hush) [inline pkg/dogstatsd/replay/state.go:41]: We should lock after this

File: pkg/dogstatsd/packets/telemetry.go (hunk 14 above)
Comment (truthbk) [reply pkg/dogstatsd/packets/telemetry.go:28]: I wanted to talk to you about this. I thought it made sense to have the telemetry counters next to the code that uses them, as opposed to have it in a separate file. If you disagree that's OK, I can move them back. This came up when I was rebasing/transplanting commits, and I decided to make that change. As said, if you prefer to have it all in one place, I'm fine with that.

File: pkg/dogstatsd/packets/buffer.go (hunk 11 above)
Comment (truthbk) [reply pkg/dogstatsd/packets/buffer.go:19]: Sure, if you feel strongly about it, I can move it back.

File: cmd/agent/api/grpc.go [hunk 18]
```diff
@@ -81,6 +83,32 @@ func (s *serverSecure) DogstatsdCaptureTrigger(ctx context.Context, req *pb.Capt
 	return &pb.CaptureTriggerResponse{Path: p}, nil
 }
 
+func (s *serverSecure) DogstatsdSetTaggerState(ctx context.Context, req *pb.TaggerState) (*pb.TaggerStateResponse, error) {
```
[the remark this answers is not in the document]
Comment (truthbk) [reply cmd/agent/api/grpc.go:86]: I'm not sure why `go vet` didn't complain about this with the pre-commit hooks? 🤔

File: pkg/dogstatsd/replay/capture.go (hunk 15 above)
Comment (truthbk) [reply pkg/dogstatsd/replay/capture.go:19]: it's just a magic number, but I can provide some context

File: cmd/agent/api/grpc.go [hunk 19]
```diff
@@ -81,6 +83,32 @@ func (s *serverSecure) DogstatsdCaptureTrigger(ctx context.Context, req *pb.Capt
...
+
+	// Reset and return if no state pushed
+	if req == nil || req.State == nil {
+		log.Debugf("API: empty request or state")
+		tagger.ResetCaptureTagger()
+		dsdReplay.SetPidMap(nil)
+		return &pb.TaggerStateResponse{Loaded: false}, nil
+	}
```
[the remark this answers is not in the document]
Comment (truthbk) [reply cmd/agent/api/grpc.go:94]: I don't think we need a dedicated reset method. An empty state submission can be interpreted like "set nil on the capture tagger", which I think is enough.

File: pkg/dogstatsd/packets/pool_manager.go (hunk 13 above)
Comment (truthbk) [reply pkg/dogstatsd/packets/pool_manager.go:64]: I _think_ type assertion has very little overhead these days. It used to be quite expensive but not anymore: https://stackoverflow.com/questions/28024884/does-a-type-assertion-type-switch-have-bad-performance-is-slow-in-go/38245293

File: pkg/dogstatsd/replay/state.go (hunk 17 above)
Comment (truthbk) [reply pkg/dogstatsd/replay/state.go:41]: Sorry, I'm not sure I follow? Why should we lock? Aren't we already holding an `RLock` there and all we're doing is reading?

File: pkg/dogstatsd/packets/pool_manager.go (hunk 13 above)
Comment (hush-hush) [reply pkg/dogstatsd/packets/pool_manager.go:64]: Indeed, but I'm not sure we need it.

File: pkg/dogstatsd/packets/telemetry.go (hunk 14 above)
Comment (hush-hush) [reply pkg/dogstatsd/packets/telemetry.go]: Some of them are used in multiple files. I felt that having everything in one place makes it easier to maintain.

File: pkg/dogstatsd/replay/state.go (hunk 17 above)
Comment (hush-hush) [reply pkg/dogstatsd/replay/state.go:39]: I meant that we should only acquire the lock after checking the `pidMap` to avoid locking if we're going to return right away.

File: pkg/dogstatsd/packets/telemetry.go (hunk 14 above)
Comment (truthbk) [reply pkg/dogstatsd/packets/telemetry.go]: I moved it all to telemetry file.

File: pkg/dogstatsd/replay/state.go (hunk 17 above)
Comment (truthbk) [reply pkg/dogstatsd/replay/state.go:39]: But the `pidMap` will be nil until the first time it's set, so it's part of the race condition. Without the `RLock()` there's a race between `pidMap` being nil and actually being something, hence why I still think we should lock before checking if it's nil.

Comment (hush-hush) [reply pkg/dogstatsd/replay/state.go:39]: my bad ... I missread the code I thought you were checking the parameter....

**Review by hush-hush — APPROVED** [review]

File: pkg/tagger/global.go [hunk 20]
```diff
@@ -61,10 +67,36 @@ func Init() {
...
+			mux.RUnlock()
+			return entity, nil
+		}
+	}
+	mux.RUnlock()
+
+	return defaultTagger.GetEntity(entityID)
+}
+
 // Tag queries the defaultTagger to get entity tags from cache or sources.
```
Comment (celenechang) [inline pkg/tagger/global.go:85]: nit suggestion: `Tag queries the captureTagger (for replay scenarios) or defaultTagger...`

File: pkg/dogstatsd/replay/state.go [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+package replay
+
+import (
+	"errors"
+	"sync"
+)
+
+var (
+	mux    sync.RWMutex
+	pidMap map[int32]string
```
Comment (xornivore) [inline pkg/dogstatsd/replay/state.go:15]: nit - maybe a type alias for the map to help with readability?

File: pkg/dogstatsd/replay/writer.go [hunk 22]
```diff
@@ -134,47 +200,52 @@ process:
...
+	if err != nil {
+		log.Errorf("There was an error flushing the underlying writer while stopping the capture: %v", err)
+	}
+
+	tc.File.Close()
 }
 
-// StopCapture stops the ongoing capture if ongoine and returns an error, if any.
+// StopCapture stops the ongoing capture if in process and returns an error, if any.
 func (tc *TrafficCaptureWriter) StopCapture() error {
```
Comment (xornivore) [inline pkg/dogstatsd/replay/writer.go:249]: error seems redundant as a return with latest changes?

File: cmd/agent/app/dogstatsd_replay.go [hunk 23]
```diff
@@ -95,16 +175,33 @@ replay:
 		select {
 		case msg := <-reader.Traffic:
 			// TODO: for when the tagger works
```
Comment (xornivore) [inline cmd/agent/app/dogstatsd_replay.go:177]: nit - this comment is stale now?

**Review by xornivore — COMMENTED** [review]
Looks great, a couple of small comments. A general question: do you envision the need to augment captured tagger state with few additional tags so that on replay it can be distinguished from similar traffic (with the purpose of A/B comparison), would that be an interesting addition in the future?

File: pkg/tagger/replay/tagger.go [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+// Init initializes the connection to the replay tagger and starts watching for
+// events.
+func (t *Tagger) Init() error {
+	t.health = health.RegisterLiveness("tagger")
+	t.telemetryTicker = time.NewTicker(1 * time.Minute)
+
+	t.ctx, t.cancel = context.WithCancel(context.Background())
+
+	// I do not think there's a need for any async work
+	// go t.run()
```
Comment (celenechang) [inline pkg/tagger/replay/tagger.go:52]: is it intentionally included?

File: pkg/tagger/replay/tagger.go [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+
+	return nil
```
Comment (celenechang) [inline pkg/tagger/replay/tagger.go:54]: i think an `error` will never be returned, maybe can rm

**Review by xornivore — APPROVED** [review]
👍

File: pkg/tagger/replay/tagger.go (hunk 25 above)
Comment (truthbk) [reply pkg/tagger/replay/tagger.go:54]: We have to satisfy the interface, so we'll have to return `nil`.
